### ISSUE TITLE
UDP-based queries timing out

### ISSUE THREAD

**Issue by atc0005** [atc0005/dnsc#17]
Example output that I'm currently seeing:

```
read udp 192.168.2.5:54418->8.8.8.8:53: i/o timeout
```

I can retry a short while later and will get this result:

```
no records found for query
```

I can't be completely certain of the reliability of the results, so going with one of these approaches seems like a good idea:

- raising the default timeout for initial query
- multiple, automatic retries with the same timeout value
- multiple, automatic retries with a slightly higher timeout for each subsequent retry

**Comment by atc0005** [comment]
> * raising the default timeout for initial query

I'm thinking that this will be the way to go, not for reliability purposes, but for judging the total time required for a query, *unless* we still also track the total time of all retries, but I think that can be worked in as a later enhancement.

Combined with the work done for #22, raising the time from what appears to be 2s to maybe 10s should allow for slower systems to report in and give a better idea which servers are slow and where there may be network connectivity issues.

**Comment by atc0005** [comment]
https://github.com/miekg/dns/blob/d128d10d176b810543f8fecd089082a29d3f159d/client.go#L16-L19

2 seconds is listed there.

Looks like it is used here:

https://github.com/miekg/dns/blob/d128d10d176b810543f8fecd089082a29d3f159d/client.go#L57-L65.

1. If the client timeout value is set, use that
1. If the connection timeout value is set, use that
1. Otherwise if neither are set, use `dnsTimeout`
    - as shown above, this is currently set to 2 seconds.

**Comment by atc0005** [comment]
https://github.com/miekg/dns/blob/d128d10d176b810543f8fecd089082a29d3f159d/client.go#L47-L55

```golang
// Exchange performs a synchronous UDP query. It sends the message m to the address
// contained in a and waits for a reply. Exchange does not retry a failed query, nor
// will it fall back to TCP in case of truncation.
// See client.Exchange for more information on setting larger buffer sizes.
func Exchange(m *Msg, a string) (r *Msg, err error) {
	client := Client{Net: "udp"}
	r, _, err = client.Exchange(m, a)
	return r, err
}
```

Basic `Client` with udp option set.

This is called:

```golang
func (c *Client) Exchange(m *Msg, address string) (r *Msg, rtt time.Duration, err error) {
	if !c.SingleInflight {
		return c.exchange(m, address)
	}

	q := m.Question[0]
	key := fmt.Sprintf("%s:%d:%d", q.Name, q.Qtype, q.Qclass)
	r, rtt, err, shared := c.group.Do(key, func() (*Msg, time.Duration, error) {
		return c.exchange(m, address)
	})
	if r != nil && shared {
		r = r.Copy()
	}

	return r, rtt, err
}
```

which ends up calling:


```golang
func (c *Client) exchange(m *Msg, a string) (r *Msg, rtt time.Duration, err error) {
	var co *Conn

	co, err = c.Dial(a)

	if err != nil {
		return nil, 0, err
	}
	defer co.Close()

	opt := m.IsEdns0()
	// If EDNS0 is used use that for size.
	if opt != nil && opt.UDPSize() >= MinMsgSize {
		co.UDPSize = opt.UDPSize()
	}
	// Otherwise use the client's configured UDP size.
	if opt == nil && c.UDPSize >= MinMsgSize {
		co.UDPSize = c.UDPSize
	}

	co.TsigSecret = c.TsigSecret
	t := time.Now()
	// write with the appropriate write timeout
	co.SetWriteDeadline(t.Add(c.getTimeoutForRequest(c.writeTimeout())))
	if err = co.WriteMsg(m); err != nil {
		return nil, 0, err
	}

	co.SetReadDeadline(time.Now().Add(c.getTimeoutForRequest(c.readTimeout())))
	r, err = co.ReadMsg()
	if err == nil && r.Id != m.Id {
		err = ErrId
	}
	rtt = time.Since(t)
	return r, rtt, err
}
```

These lines appear relevant:

https://github.com/miekg/dns/blob/d128d10d176b810543f8fecd089082a29d3f159d/client.go#L167

```golang
	co.SetWriteDeadline(t.Add(c.getTimeoutForRequest(c.writeTimeout())))
```

https://github.com/miekg/dns/blob/d128d10d176b810543f8fecd089082a29d3f159d/client.go#L172

```golang
	co.SetReadDeadline(time.Now().Add(c.getTimeoutForRequest(c.readTimeout())))
```

All of this to say, the default 2s is used if another value isn't used to override. Since the `Client` type appears to be where the values are pulled (by way of the default if nothing else), we need to focus there.

**Comment by atc0005** [comment]
Recap for my future set:

1. Create a custom `Client`
1. Set value on custom `Client`

The other bits are adding support for providing the timeout value via config file and flag. As noted on #10 overriding default values via config file is currently broken (intending to fix in #10), but we can go ahead and get the support in for setting the value in just one of the two places.
